examples/aws/terraform/ha-autoscale-cluster/README.md

```diff
@@ -67,6 +66,13 @@ export TF_VAR_route53_zone="example.com"
 # This will be used for internet access for users connecting to teleport proxy
 export TF_VAR_route53_domain="cluster.example.com"
 
+# (optional) If using ACM, set an additional DNS alias which will be added pointing to the NLB. This can
+# be used with clients like kubectl which should target a DNS record. This will also add the DNS name to the
+# Teleport Kubernetes config to prevent certificate SNI issues. You can use this DNS name with commands like:
+# `tctl auth sign --user=foo --format=kubernetes --out=kubeconfig --proxy=https://cluster-nlb.example.com:3026`
+# This setting only takes effect when using ACM, it will be ignored otherwise.
+#export TF_VAR_route53_domain_acm_nlb_alias="cluster-nlb.example.com"
```

Comment:
instead of an additional NLB alias, can this be called `TF_VAR_route53_domain_acm_kube_domain`?
basically an override of `kubernetes.publc_addr`

Reply:
I'm not sure I follow; is that just a change in naming? I think it already does everything else you suggested.

Reply:
It would also mean changing:
```
        if [[ "${TELEPORT_PROXY_SERVER_NLB_ALIAS}" != "" ]]; then
            echo "    public_addr: ['${TELEPORT_PROXY_SERVER_LB}:3026', '${TELEPORT_PROXY_SERVER_NLB_ALIAS}:3026']" >> /etc/teleport.yaml
        else
            echo "    public_addr: ['${TELEPORT_PROXY_SERVER_LB}:3026']" >> /etc/teleport.yaml
        fi
```
to
```
        if [[ "${TELEPORT_PROXY_SERVER_KUBE_DOMAIN}" != "" ]]; then
            echo "    public_addr: ['${TELEPORT_PROXY_SERVER_KUBE_DOMAIN}:3026']" >> /etc/teleport.yaml
        else
            echo "    public_addr: ['${TELEPORT_PROXY_SERVER_LB}:3026']" >> /etc/teleport.yaml
        fi
```

Reply:
Ah, I see what you mean. Yes, we could make it the only allowable hostname for accessing the kube proxy - is the reason you want to do this because it would automatically get added to any certs being issues without needing to use the `--proxy` flag to `tctl auth sign`?

I suppose it's quite unlikely that anyone setting an alias would want to use the raw NLB address too.

Reply:
If a user has a separate domain name for k8s access, there's probably a reason they can't go through the main proxy domain. Putting *only* the custom k8s domain in the cert makes it harder to mis-configure and route some traffic around the correct LB.

But from Teleport's perspective, there isn't much downside in putting multiple domains in the cert. If you think there's value in that, feel free to keep the existing logic.

Regardless, the naming could be improved. `TELEPORT_PROXY_SERVER_NLB_ALIAS` is non-obvious and also suggests that this address is used for other non-k8s endpoints.

Reply:
Not sure if it was 100% clear but the reason we're adding the separate domain here is only because if you use ACM with a domain like `teleport.example.com`, it configures two load balancers:
- an ALB with an HTTPS listener to serve `teleport.example.com` (mandatory, as AWS holds the private key for the cert)
- a separate NLB for all the TCP listeners (including Kubernetes)

We set the NLB's default DNS name (something like `gus-tftestkube4-proxy-bc9ba568645c3d80.elb.us-east-1.amazonaws.com`) as the `public_addr` for all the TCP-based services in Teleport's config file, then `tsh` automatically discovers this name via API. `kubectl` can't do that (at least with a static `kubeconfig` from `tctl auth sign`) because it doesn't have any way to query the Teleport API.

Come to think of it though, we probably ought to configure/allow the use of an NLB alias like this for all Teleport's TCP-based services when using ACM. One issue with the current method is that if the NLB's DNS name changes for any reason, Teleport's config doesn't get updated and so connections will fail until proxy instances are cycled via ASG. Using the alias would prevent this.

Reply:
Sounds like a real Rube Goldberg machine :sweat_smile: 

The address used by `kubectl` is injected by `tsh` and should be coming from [the same address discovery API](https://github.com/gravitational/teleport/blob/fde61d71e0b88dcbdb00a1b5a36a9a55e553d123/lib/client/weblogin.go#L150-L159) as with SSH. Maybe `tctl auth sign` doesn't follow that logic?

For now, this PR should be fine, as long as all affected endpoints are covered.